apollo-mockserver/src/jvmMain/kotlin/com/apollographql/apollo3/mockserver/MockServer.kt

```diff
@@ -1,65 +1,80 @@
 package com.apollographql.apollo3.mockserver
 
-import okhttp3.Headers
-import okhttp3.mockwebserver.Dispatcher
-import okhttp3.mockwebserver.MockWebServer
-import okhttp3.mockwebserver.RecordedRequest
-import okio.Buffer
-import java.util.concurrent.TimeUnit
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import okio.buffer
+import okio.sink
+import okio.source
+import java.net.ServerSocket
+import java.net.Socket
+import java.util.concurrent.Executors
 
-actual class MockServer actual constructor(override val mockServerHandler: MockServerHandler) : MockServerInterface {
-  private val mockWebServer = MockWebServer().apply { dispatcher = mockServerHandler.toOkHttpDispatcher() }
+@Suppress("BlockingMethodInNonBlockingContext")
+actual class MockServer actual constructor(
+    override val mockServerHandler: MockServerHandler,
+) : MockServerInterface {
+  private val serverSocket = ServerSocket(0)
+  private val mockRequests = mutableListOf<MockRequest>()
 
-  override fun enqueue(mockResponse: MockResponse) {
-    (mockServerHandler as? QueueMockServerHandler)?.enqueue(mockResponse)
-        ?: error("Apollo: cannot call MockServer.enqueue() with a custom handler")
-  }
+  private val dispatcher = Executors.newFixedThreadPool(5).asCoroutineDispatcher()
+  private val coroutineScope = CoroutineScope(SupervisorJob() + dispatcher)
 
-  override fun takeRequest(): MockRequest {
-    return mockWebServer.takeRequest(10, TimeUnit.MILLISECONDS)?.toApolloMockRequest() ?: error("No recorded request")
+  init {
+    coroutineScope.launch {
+      while (true) {
+        val clientSocket = try {
+          serverSocket.accept()
+        } catch (_: Exception) {
+          // An exception here means the server socket has been closed (stop() was called)
+          break
+        }
+        coroutineScope.launch { handleClient(clientSocket) }
```

Comment:
Use the receiver scope to honor structured concurrency?

```suggestion
       launch { handleClient(clientSocket) }
```